File: AI_SKILL_LIBRARY/tests/test_memory_lifecycle.py
from __future__ import annotations

import unittest

from AI_SKILL_LIBRARY.v4.tools.memory_lifecycle import evaluate_candidate, transition


BASE = {
    "candidate_id": "m1",
    "domain": "engineering",
    "scope": "project:ai_brain",
    "content": "Universal Entry uses one canonical Brain authority.",
    "source": "verified_test",
    "confidence": 0.9,
    "created_at": "2026-09-16T03:00:00Z",
    "evidence_refs": ["spec:universal-brain-fabric"],
    "reusable": True,
    "verified": True,
    "non_sensitive": True,
}


class MemoryLifecycleTests(unittest.TestCase):
    def test_candidate_never_promotes_directly(self) -> None:
        out = evaluate_candidate(dict(BASE), [])
        self.assertTrue(out["eligible_for_review"])
        self.assertEqual(out["state"], "candidate")
        self.assertFalse(out["active"])

    def test_missing_evidence_holds_candidate(self) -> None:
        row = dict(BASE, evidence_refs=[])
        out = evaluate_candidate(row, [])
        self.assertFalse(out["eligible_for_review"])
        self.assertIn("missing_evidence", out["reasons"])

    def test_conflict_blocks_review(self) -> None:
        active = [{
            "memory_id": "old",
            "domain": "engineering",
            "scope": "project:ai_brain",
            "content": "Universal Entry uses a parallel Brain authority.",
            "state": "active",
            "verified": True,
            "created_at": "2026-09-15T03:00:00Z",
        }]
        row = dict(BASE, conflicts_with=["old"])
        out = evaluate_candidate(row, active)
        self.assertFalse(out["eligible_for_review"])
        self.assertIn("unresolved_conflict", out["reasons"])

    def test_secret_and_raw_chat_are_rejected(self) -> None:
        row = dict(BASE, content="api_key=SECRET123")
        out = evaluate_candidate(row, [])
        self.assertFalse(out["eligible_for_review"])
        self.assertIn("sensitive_content", out["reasons"])
        raw = dict(BASE)
        raw["raw_private_chat"] = "private transcript"
        out = evaluate_candidate(raw, [])
        self.assertFalse(out["eligible_for_review"])
        self.assertIn("forbidden_field", out["reasons"])

    def test_active_transition_requires_verified_gate(self) -> None:
        candidate = dict(BASE, state="candidate")
        with self.assertRaises(ValueError):
            transition(candidate, "active", {"verified": False})
        active = transition(candidate, "active", {
            "verified": True,
            "scoped": True,
            "reusable": True,
            "non_sensitive": True,
            "conflict_free": True,
            "verified_at": "2026-09-16T04:00:00Z",
        })
        self.assertEqual(active["state"], "active")
        self.assertEqual(active["last_verified"], "2026-09-16T04:00:00Z")

    def test_verified_newer_can_supersede(self) -> None:
        active = transition(dict(BASE, state="candidate"), "active", {
            "verified": True,
            "scoped": True,
            "reusable": True,
            "non_sensitive": True,
            "conflict_free": True,
            "verified_at": "2026-09-16T04:00:00Z",
        })
        superseded = transition(active, "superseded", {
            "verified": True,
            "superseded_by": "m2",
            "verified_at": "2026-09-16T05:00:00Z",
        })
        self.assertEqual(superseded["state"], "superseded")
        self.assertEqual(superseded["superseded_by"], "m2")


if __name__ == "__main__":
    unittest.main()

File: cloudflare-worker/test-memory-lifecycle.mjs
import assert from 'node:assert/strict';
import {isSensitiveMemoryContent,promoteMemoryCandidate,reviewMemoryCandidate,supersedeMemory} from './memory-lifecycle.js';

const candidate={
  candidate_id:'m-123',domain:'coding',scope:'project:trading-api',content:'Universal Fabric uses one canonical Brain authority.',source:'verified-task',confidence:0.9,created_at:'2026-09-16T00:00:00Z',evidence_refs:['eval:1'],reusable:true,verified:true,non_sensitive:true,conflicts_with:[],submitted_by:'chatgpt',state:'candidate',active:false,
};

const review=reviewMemoryCandidate(candidate,{evidence_count:2,conflict:false,current:true,reviewed_at:'2026-09-16T01:00:00Z'});
assert.equal(review.decision,'confirmed');
const active=promoteMemoryCandidate(candidate,review);
assert.equal(active.state,'active');
assert.equal(active.active,true);
assert.equal(active.last_verified,'2026-09-16T01:00:00Z');
assert.deepEqual(active.provenance.evidence_refs,['eval:1']);

assert.throws(()=>promoteMemoryCandidate({...candidate,non_sensitive:false},review),/memory_promotion_blocked/);
assert.throws(()=>promoteMemoryCandidate({...candidate,confidence:0.54},review),/memory_promotion_blocked/);
assert.equal(reviewMemoryCandidate({...candidate,conflicts_with:['m-old']},{evidence_count:2,current:true,conflict:false}).decision,'rejected');
assert.equal(reviewMemoryCandidate(candidate,{evidence_count:0,current:true,conflict:false}).decision,'needs_reverify');
assert.equal(reviewMemoryCandidate(candidate,{evidence_count:1,current:true,conflict:true}).decision,'rejected');

const superseded=supersedeMemory(active,{memory_id:'m-124',last_verified:'2026-09-16T02:00:00Z'});
assert.equal(superseded.state,'superseded');
assert.equal(superseded.active,false);
assert.equal(superseded.superseded_by,'m-124');
assert.deepEqual(superseded.provenance,active.provenance);

const secret={...candidate,content:'api_key = top-secret-value'};
assert.equal(isSensitiveMemoryContent(secret),true);
assert.equal(reviewMemoryCandidate(secret,{evidence_count:10,current:true,conflict:false}).decision,'rejected');
assert.throws(()=>promoteMemoryCandidate(secret,{decision:'confirmed',evidence_count:10,current:true,conflict:false}),/memory_promotion_blocked/);
const rawChat={...candidate,raw_private_chat:'private text'};
assert.equal(isSensitiveMemoryContent(rawChat),true);
assert.equal(reviewMemoryCandidate(rawChat,{evidence_count:10,current:true,conflict:false}).decision,'rejected');

console.log('MEMORY_LIFECYCLE_TESTS=PASS');
